### PR TITLE
Feature: Progressive Messages

### PR DESCRIPTION
**Pull request by benrr101**
This change is a reworking of the way that messages are sent to clients from the service layer. It is also a reworking of the protocol to ensure that all formulations of query send back events to the client in a deterministic ordering. To support the first change:
* Added a new event that will be sent when a message is generated
* Messages now indicate which Batch (if any) generated them
* Messages now indicate if they were error level
* Removed message storage in Batch objects and BatchSummary objects
* Batch objects no longer have error state

### CODE REVIEW

**Comment by kburtram** [comment]
Please make sure the AppVeyor is clean prior to merging.

**Comment by benrr101** [comment]
Of course, working on that right now

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Batch.cs [hunk 1]
```diff
@@ -84,11 +83,22 @@ internal Batch(string batchText, SelectionData selection, int ordinalId, IFileSt
...
+
+        /// <summary>
         /// Event that will be called when the batch has completed execution
         /// </summary>
         public event BatchAsyncEventHandler BatchCompletion;
 
         /// <summary>
+        /// Event that will be called when a message has been emitted
+        /// </summary>
+        public event BatchAsyncMessageHandler BatchMessage;
```
Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Batch.cs]: I think the better name would be BatchMessageSent, BatchMessage doesn't really say why happened to the message.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Batch.cs [hunk 2]
```diff
@@ -430,30 +429,30 @@ internal void StatementCompletedHandler(object sender, StatementCompletedEventAr
         /// </summary>
         /// <param name="sender">Object that fired the event</param>
         /// <param name="args">Arguments from the event</param>
-        private void StoreDbMessage(object sender, SqlInfoMessageEventArgs args)
+        private void ServerMessageHandler(object sender, SqlInfoMessageEventArgs args)
         {
-            resultMessages.Add(new ResultMessage(args.Message));
+            SendMessage(args.Message, false).Wait();
```
Comment (llali) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Batch.cs]: sendMessage returns null if the message is null, should you check for null here before running .Wait()?

Comment (benrr101) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Batch.cs]: I don't think that's the case. SendMessage is a async method, so it always returns a Task.

**Comment by benrr101** [comment]
@kburtram Coveralls shows unknown changes... Is that good or no good?

**Comment by llali** [comment]
:shipit:

**Comment by benrr101** [comment]
Code coverage is approximately the same as it was before (according to a manually ran Jenkins build). I'll go ahead and check this in.

**Pull request merged**